lib/HLSL/DxilScalarizeVectorLoadStores.cpp

```diff
@@ -0,0 +1,239 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilScalarizeVectorLoadStores.cpp                                         //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Lowers native vector load stores to potentially multiple scalar calls.    //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/DXIL/DxilInstructions.h"
+#include "dxc/DXIL/DxilModule.h"
+#include "dxc/HLSL/DxilGenerationPass.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+class DxilScalarizeVectorLoadStores : public ModulePass {
+private:
+  DxilModule *m_DM;
+
+  void scalarizeVectorLoad(hlsl::OP *HlslOP, const DataLayout &DL,
+                           CallInst *CI);
+  void scalarizeVectorStore(hlsl::OP *HlslOP, const DataLayout &DL,
+                            CallInst *CI);
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilScalarizeVectorLoadStores() : ModulePass(ID) {}
+
+  StringRef getPassName() const override {
+    return "DXIL scalarize vector load/stores";
+  }
+
+  bool runOnModule(Module &M) override {
+    DxilModule &DM = M.GetOrCreateDxilModule();
+    m_DM = &DM;
+
+    // Shader Model 6.9 allows native vectors and doesn't need this pass.
+    if (DM.GetShaderModel()->IsSM69Plus())
+      return false;
+
+    bool Changed = false;
+
+    hlsl::OP *HlslOP = DM.GetOP();
+    for (auto F = M.functions().begin(), E = M.functions().end(); F != E;) {
+      Function *Func = &*(F++);
+      DXIL::OpCodeClass OpClass;
+      if (HlslOP->GetOpCodeClass(Func, OpClass)) {
+        if (OpClass == DXIL::OpCodeClass::RawBufferVectorLoad) {
+          for (auto U = Func->user_begin(), UE = Func->user_end(); U != UE;) {
+            CallInst *CI = cast<CallInst>(*(U++));
+            scalarizeVectorLoad(HlslOP, M.getDataLayout(), CI);
```

Comment:
It's always a good form when iterating DXIL op class overload function users to check the opcode (even though you know in this case it should be the only one for this OpCodeClass):
```suggestion
            if (HlslOP->GetDxilOpFuncCallInst(CI) !=
                hlsl::OP::OpCode::RawBufferVectorLoad)
              continue;
            scalarizeVectorLoad(HlslOP, M.getDataLayout(), CI);
```

Reply:
It turns out that you *do* technically have to check, even with `GetOpFuncList`.  That's because this returns the list of overload functions for the `OpCodeClass` corresponding to the `OpCode`.  Basically, that's all it could do, since different DXIL operations use the same function declaration when they share the same `OpCodeClass`.

I think the best way is to get the `OpCode` using: `hlsl::OP::GetDxilOpFuncCallInst(CI) == hlsl::OP::OpCode::RawBufferVectorLoad` (this avoids a redundant `IsDxilOpFuncCallInst` check on `NDEBUG` at least).

I looked at other code using this function and basically *nobody* is checking the `OpCode` of the `CallInst` users.  I'm considering filing an issue about this, but I don't think there are any real bugs caused by this at the moment.  Basically, you're safe if there's a 1:1 mapping between `OpCode` and `OpCodeClass`, which happens to be true in all use cases.  If not, since this always lists all the overloads for the `OpCodeClass` that corresponds to the `OpCode` passed in, you could be iterating `CallInst` users that aren't calling with the same `OpCode` you're looking for.  I found one use of `GetOpFuncList` that looks like it could potentially have this problem in `OP::IsDxilOpUsed`, but this is only called with `OpCode`s that are safe, so I don't think it doesn't cause any real bugs in practice (yet).

I think it would probably be better to make `GetOpFuncList` take the `OpCodeClass` instead of the `OpCode`.  Then you're forced to `HlslOp->GetOpFuncList(hlsl::OP::GetOpCodeClass(MyOpCode))` and it's more obvious that you need to check the `OpCode` for each `CallInst` user.